src/raftstore/coprocessor/dispatcher.rs

```diff
@@ -16,16 +18,91 @@ struct Entry<T> {
     observer: T,
 }
 
-// TODO: change it to Send + Clone.
-pub type BoxAdminObserver = Box<dyn AdminObserver + Send + Sync>;
-pub type BoxQueryObserver = Box<dyn QueryObserver + Send + Sync>;
-pub type BoxApplySnapshotObserver = Box<dyn ApplySnapshotObserver + Send + Sync>;
-pub type BoxSplitCheckObserver = Box<dyn SplitCheckObserver + Send + Sync>;
-pub type BoxRoleObserver = Box<dyn RoleObserver + Send + Sync>;
-pub type BoxRegionChangeObserver = Box<dyn RegionChangeObserver + Send + Sync>;
+impl<T: Clone> Clone for Entry<T> {
+    fn clone(&self) -> Self {
+        Self {
+            priority: self.priority,
+            observer: self.observer.clone(),
+        }
+    }
+}
+
+pub trait ClonableObserver: 'static + Send {
+    type Ob: ?Sized + Send;
+    fn inner(&self) -> &Self::Ob;
+    fn inner_mut(&mut self) -> &mut Self::Ob;
+    fn box_clone(&self) -> Box<dyn ClonableObserver<Ob = Self::Ob> + Send>;
+}
+
+macro_rules! impl_box_observer {
+    ($name:ident, $ob: ident, $wrapper: ident) => {
+        pub struct $name(Box<dyn ClonableObserver<Ob = dyn $ob> + Send>);
+        impl $name {
+            pub fn new<T: 'static + $ob + Clone>(observer: T) -> $name {
+                $name(Box::new($wrapper { inner: observer }))
+            }
+        }
+        impl Clone for $name {
+            fn clone(&self) -> $name {
+                $name((**self).box_clone())
+            }
+        }
+        impl Deref for $name {
+            type Target = Box<dyn ClonableObserver<Ob = dyn $ob> + Send>;
+
+            fn deref(&self) -> &Box<dyn ClonableObserver<Ob = dyn $ob> + Send> {
+                &self.0
+            }
+        }
+        impl DerefMut for $name {
+            fn deref_mut(&mut self) -> &mut Box<dyn ClonableObserver<Ob = dyn $ob> + Send> {
+                &mut self.0
+            }
+        }
```

Comment:
Where is this used? Is there any code that changes the observer's state? If so, that would be a problem, since it only mutates only one clone of the observer, and other clones of the observer won't be mutated.

Reply:
Not yet. Shall we remove this?

Reply:
I think it might be better not to allow mutating it, so that in case there are multiple clones of the observer, their states will not be inconsistent.